fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMutuallyByRecords.java

```diff
@@ -0,0 +1,529 @@
+         */
+        if (fragmentIterationType == FragmentIterationType.RECOVER) {
+            // At this point, recovery should be done manually after investigation. In the future we may fallback
+            // to a regular multi target indexing - with or without a type stamp change.
+            throw new ValidationException("Mutual indexing failure - third iteration");
+        }
+        validateSameMetadataOrThrow(store);
+        RangeSet rangeSet = new RangeSet(store.indexRangeSubspace(common.getPrimaryIndex()));
+        return rangeSet.missingRanges(store.ensureContextActive()).asList().thenCompose(missingRanges ->
+                buildNextRangeOnly(sortAndSquash(missingRanges), subspaceProvider, subspace));
+    }
+
+    private CompletableFuture<Boolean> buildNextRangeOnly(List<Range> missingRanges,
+                                                          @Nonnull SubspaceProvider subspaceProvider, @Nonnull Subspace subspace) {
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug(KeyValueLogMessage.of("buildNextRangeOnly",
+                    LogMessageKeys.MISSING_RANGES, missingRanges));
+        }
+        if (missingRanges.isEmpty()) {
+            return AsyncUtil.READY_FALSE; // all done, Hurray!
+        }
+        // Here:
+        // for each fragmented step, if it should be built - build it and return true
+        while (true) {
+            final Range fragmentRange = fragmentGet();
+            if (fragmentIterationType == FragmentIterationType.RECOVER) {
+                if (LOGGER.isWarnEnabled()) {
+                    LOGGER.warn(KeyValueLogMessage.of("Entering recovery mode",
+                            LogMessageKeys.SPLIT_RANGES, missingRanges));
+                }
+                return AsyncUtil.READY_TRUE; // let the caller handle the recovery
+            }
+            Range rangeToBuild = fragmentIterationType == FragmentIterationType.FULL ?
+                                 fullyUnBuiltRange(missingRanges, fragmentRange) :
+                                 partlyUnBuiltRange(missingRanges, fragmentRange);
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug(KeyValueLogMessage.of("range to build",
+                        LogMessageKeys.SCAN_TYPE, fragmentIterationType,
+                        LogMessageKeys.RANGE, rangeToBuild,
+                        LogMessageKeys.ORIGINAL_RANGE, fragmentRange));
+            }
+            if (rangeToBuild != null) {
+                infiniteLoopProtection(rangeToBuild, missingRanges);
+                return iterateAllRanges(null,
+                        (store, recordsScanned) -> buildThisRangeOnly(store, recordsScanned, rangeToBuild),
+                        subspaceProvider, subspace).thenCompose(ignore -> AsyncUtil.READY_TRUE);
+            }
+            fragmentPlusPlus();
+        }
+    }
+
+    private CompletableFuture<Boolean> buildThisRangeOnly(@Nonnull FDBRecordStore store, @Nonnull AtomicLong recordsScanned,
```

Comment:
This code is almost exactly the same `IndexingMultiTargetByRecords`, can more code be reused between the two? I saw some other duplication, but that was much smaller and less delicate.

Reply:
In the first draft `IndexingMutuallyByRecords` had inherited `IndexingMultiTargetByRecords`. The slight changes between `buildThisRangeOnly` and `buildRangeOnly` were enough to make the code *very* hard to read & maintain on both modules. Despite the similarity, I am much happier with this version.